homeassistant/components/sms/__init__.py

```diff
@@ -17,17 +20,53 @@
 )
 
 
-def setup(hass, config):
+async def async_setup(hass, config):
     """Configure Gammu state machine."""
-    conf = config[DOMAIN]
-    device = conf.get(CONF_DEVICE)
-    gateway = gammu.StateMachine()  # pylint: disable=no-member
-    try:
-        gateway.SetConfig(0, dict(Device=device, Connection="at"))
-        gateway.Init()
-    except gammu.GSMError as exc:  # pylint: disable=no-member
-        _LOGGER.error("Failed to initialize, error %s", exc)
+    hass.data.setdefault(DOMAIN, {})
+    sms_config = config.get(DOMAIN, {})
+    hass.data[DOMAIN][NOTIFY_DOMAIN] = config
+    if not sms_config:
+        return True
+
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": SOURCE_IMPORT}, data=sms_config,
+        )
+    )
+
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Configure Gammu state machine."""
+
+    device = entry.data[CONF_DEVICE]
+    config = {"Device": device, "Connection": "at"}
+    gateway = await create_sms_gateway(config, hass)
+    if not gateway:
         return False
-    else:
-        hass.data[DOMAIN] = gateway
+    hass.data[DOMAIN][SMS_GATEWAY] = gateway
+
+    # Notify doesn't support config entry setup yet, load with discovery for now
+    hass_config = hass.data[DOMAIN][NOTIFY_DOMAIN]
+    notify_configs = hass.data[DOMAIN][NOTIFY_DOMAIN].get(NOTIFY_DOMAIN, {})
+    for notify_config in notify_configs:
+        hass.async_create_task(
+            discovery.async_load_platform(
+                hass, NOTIFY_DOMAIN, DOMAIN, notify_config, hass_config,
+            )
+        )
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Unload a config entry."""
+
+    if SMS_GATEWAY in hass.data[DOMAIN]:
+        gateway = hass.data[DOMAIN][SMS_GATEWAY]
+        hass.data[DOMAIN].pop(SMS_GATEWAY)
+        await gateway.terminate_async()
         return True
+
+    return False
```

Comment:
We don't need this check. We know the `SMS_GATEWAY` will be in hass.data since the entry must be setup before it's unloaded.
```suggestion
    gateway = hass.data[DOMAIN].pop(SMS_GATEWAY)
    await gateway.terminate_async()
    return True
```